{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#https://cooltools.readthedocs.io/en/latest/notebooks/viz.html\n",
    "\n",
    "#Import the packages we will use\n",
    "#Utilities\n",
    "import os\n",
    "import re\n",
    "import itertools\n",
    "from itertools import combinations\n",
    "import glob\n",
    "import pickle\n",
    "import argparse\n",
    "\n",
    "#Data Management\n",
    "import numpy as np\n",
    "from numpy import diff\n",
    "import pandas as pd\n",
    "import h5py\n",
    "import scipy\n",
    "from scipy.stats import linregress\n",
    "from scipy import ndimage\n",
    "from functools import partial\n",
    "from scipy.linalg import toeplitz\n",
    "\n",
    "#Plotting\n",
    "%matplotlib inline\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "from matplotlib import cm\n",
    "from matplotlib.gridspec import GridSpec\n",
    "from matplotlib.gridspec import GridSpecFromSubplotSpec\n",
    "import matplotlib.colors as colors\n",
    "from matplotlib.colors import ListedColormap\n",
    "import seaborn as sns\n",
    "import upsetplot\n",
    "from upsetplot import UpSet\n",
    "\n",
    "#Genomics\n",
    "import pairtools\n",
    "import cooler\n",
    "import cooltools\n",
    "# Import python package for working with cooler files and tools for analysis\n",
    "import cooler\n",
    "import cooltools.lib.plotting\n",
    "import bioframe\n",
    "from bioframe import overlap\n",
    "import bbi\n",
    "from cooltools import insulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Using hg38 aligned files, sampling coolers to same number of interactions between samples/replicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "outDataDir = '..'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#conditions\n",
    "conditions = [\n",
    "    'WT_Ctrl_R1',\n",
    "    'WT_ATRA_R1',\n",
    "    'Top2BKO_Ctrl_R1',\n",
    "    'Top2BKO_ATRA_R1',\n",
    "    'WT_Ctrl_R2',\n",
    "    'WT_ATRA_R2',\n",
    "    'Top2BKO_Ctrl_R2',\n",
    "    'Top2BKO_ATRA_R2',\n",
    "    'WT_Ctrl_R1R2',\n",
    "    'WT_ATRA_R1R2',\n",
    "    'Top2BKO_Ctrl_R1R2',\n",
    "    'Top2BKO_ATRA_R1R2'\n",
    "]\n",
    "\n",
    "long_names = {\n",
    "    'WT_Ctrl_R1' : 'CA-HiC-Dpn-SH-SY5Y-WT-Ctrl-4-51-R1-T1',\n",
    "    'WT_ATRA_R1' : 'CA-HiC-Dpn-SH-SY5Y-WT-ATRA-5days-4-51-R1-T1',\n",
    "    'Top2BKO_Ctrl_R1' : 'CA-HiC-Dpn-SH-SY5Y-BKO98-Ctrl-4-51-R1-T1',\n",
    "    'Top2BKO_ATRA_R1' : 'CA-HiC-Dpn-SH-SY5Y-BKO98-ATRA-5days-4-51-R1-T1',\n",
    "    'WT_Ctrl_R2' : 'CA-HiC-Dpn-SH-SY5Y-WT-Ctrl-4-52-R2-T1',\n",
    "    'WT_ATRA_R2' : 'CA-HiC-Dpn-SH-SY5Y-WT-ATRA-5days-4-52-R2-T1',\n",
    "    'Top2BKO_Ctrl_R2' : 'CA-HiC-Dpn-SH-SY5Y-BKO98-Ctrl-4-52-R2-T1',\n",
    "    'Top2BKO_ATRA_R2' : 'CA-HiC-Dpn-SH-SY5Y-BKO98-ATRA-5days-4-52-R2-T1',\n",
    "    'WT_Ctrl_R1R2' : 'CA-HiC-Dpn-SH-SY5Y-WT-Ctrl-4-51-and-4-52-R1R2',\n",
    "    'WT_ATRA_R1R2' : 'CA-HiC-Dpn-SH-SY5Y-WT-ATRA-5days-4-51-and-4-52-R1R2',\n",
    "    'Top2BKO_Ctrl_R1R2' : 'CA-HiC-Dpn-SH-SY5Y-BKO98-Ctrl-4-51-and-4-52-R1R2',\n",
    "    'Top2BKO_ATRA_R1R2' : 'CA-HiC-Dpn-SH-SY5Y-BKO98-ATRA-5days-4-51-and-4-52-R1R2'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#add colors for each sample\n",
    "sampleColors = {\n",
    "    'WT_Ctrl_R1' : '#a6cee3',\n",
    "    'WT_ATRA_R1' : '#1f78b4',\n",
    "    'Top2BKO_Ctrl_R1' : '#b2df8a',\n",
    "    'Top2BKO_ATRA_R1' : '#33a02c',\n",
    "    'WT_Ctrl_R2' : '#a6cee3',\n",
    "    'WT_ATRA_R2' : '#1f78b4',\n",
    "    'Top2BKO_Ctrl_R2' : '#b2df8a',\n",
    "    'Top2BKO_ATRA_R2' : '#33a02c',\n",
    "    'WT_Ctrl_R1R2' : '#a6cee3',\n",
    "    'WT_ATRA_R1R2' : '#1f78b4',\n",
    "    'Top2BKO_Ctrl_R1R2' : '#b2df8a',\n",
    "    'Top2BKO_ATRA_R1R2' : '#33a02c'\n",
    "}\n",
    "\n",
    "sampleLineStyles = {\n",
    "    'WT_Ctrl_R1' : '--',\n",
    "    'WT_ATRA_R1' : '--',\n",
    "    'Top2BKO_Ctrl_R1' : '--',\n",
    "    'Top2BKO_ATRA_R1' : '--',\n",
    "    'WT_Ctrl_R2' : ':',\n",
    "    'WT_ATRA_R2' : ':',\n",
    "    'Top2BKO_Ctrl_R2' : ':',\n",
    "    'Top2BKO_ATRA_R2' : ':',\n",
    "    'WT_Ctrl_R1R2' : '-',\n",
    "    'WT_ATRA_R1R2' : '-',\n",
    "    'Top2BKO_Ctrl_R1R2' : '-',\n",
    "    'Top2BKO_ATRA_R1R2' : '-'\n",
    "}\n",
    "\n",
    "samplePlotNames = {\n",
    "    'WT_Ctrl_R1' : 'WT Ctrl, R1',\n",
    "    'WT_ATRA_R1' : 'WT ATRA, R1',\n",
    "    'Top2BKO_Ctrl_R1' : 'BKO Ctrl, R1',\n",
    "    'Top2BKO_ATRA_R1' : 'BKO ATRA, R1',\n",
    "    'WT_Ctrl_R2' : 'WT Ctrl, R2',\n",
    "    'WT_ATRA_R2' : 'WT ATRA, R2',\n",
    "    'Top2BKO_Ctrl_R2' : 'BKO Ctrl, R2',\n",
    "    'Top2BKO_ATRA_R2' : 'BKO ATRA, R2',\n",
    "    'WT_Ctrl_R1R2' : 'WT Ctrl',\n",
    "    'WT_ATRA_R1R2' : 'WT ATRA',\n",
    "    'Top2BKO_Ctrl_R1R2' : 'BKO Ctrl',\n",
    "    'Top2BKO_ATRA_R1R2' : 'BKO ATRA'    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SepConds = [\n",
    "    'WT_Ctrl_R1',\n",
    "    'WT_ATRA_R1',\n",
    "    'Top2BKO_Ctrl_R1',\n",
    "    'Top2BKO_ATRA_R1',\n",
    "    'WT_Ctrl_R2',\n",
    "    'WT_ATRA_R2',\n",
    "    'Top2BKO_Ctrl_R2',\n",
    "    'Top2BKO_ATRA_R2',\n",
    "]\n",
    "\n",
    "ComboConds = [\n",
    "    'WT_Ctrl_R1R2',\n",
    "    'WT_ATRA_R1R2',\n",
    "    'Top2BKO_Ctrl_R1R2',\n",
    "    'Top2BKO_ATRA_R1R2'    \n",
    "]\n",
    "\n",
    "SepCtrlConds = [\n",
    "    'WT_Ctrl_R1',\n",
    "    'WT_Ctrl_R1',\n",
    "    'WT_ATRA_R1',\n",
    "    'Top2BKO_Ctrl_R1',\n",
    "    'WT_Ctrl_R2',\n",
    "    'WT_Ctrl_R2',\n",
    "    'WT_ATRA_R2',\n",
    "    'Top2BKO_Ctrl_R2',\n",
    "]\n",
    "\n",
    "SepTreatConds = [\n",
    "    'WT_ATRA_R1',\n",
    "    'Top2BKO_Ctrl_R1',\n",
    "    'Top2BKO_ATRA_R1',\n",
    "    'Top2BKO_ATRA_R1',\n",
    "    'WT_ATRA_R2',\n",
    "    'Top2BKO_Ctrl_R2',\n",
    "    'Top2BKO_ATRA_R2',\n",
    "    'Top2BKO_ATRA_R2',   \n",
    "]\n",
    "\n",
    "ComboCtrlConds = [\n",
    "    'WT_Ctrl_R1R2',\n",
    "    'WT_Ctrl_R1R2',\n",
    "    'WT_ATRA_R1R2',\n",
    "    'Top2BKO_Ctrl_R1R2'\n",
    "]\n",
    "\n",
    "ComboTreatConds = [\n",
    "    'WT_ATRA_R1R2',\n",
    "    'Top2BKO_Ctrl_R1R2',\n",
    "    'Top2BKO_ATRA_R1R2',\n",
    "    'Top2BKO_ATRA_R1R2'  \n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#multi-chromosome heatmaps, 5Mb bins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#coolers - 5Mb bins\n",
    "binsize = 5000000\n",
    "\n",
    "clr_paths_5Mb = {}\n",
    "for cond in conditions:\n",
    "    clr_paths_5Mb[cond] = f'{outDataDir}/data/{long_names[cond]}.sampled.hg38.mapq_30.1000.mcool::resolutions/{binsize}'\n",
    "    \n",
    "clrs5Mb = {\n",
    "    cond: cooler.Cooler(clr_paths_5Mb[cond]) for cond in conditions\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cooltools.lib.plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ComboConds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"ticks\")\n",
    "sns.set_context(\"paper\")\n",
    "\n",
    "for cond in ComboConds:\n",
    "    \n",
    "    binsize = 5000000\n",
    "\n",
    "    opts = dict(\n",
    "        vmin=-3.5, #change scale range here\n",
    "        vmax=-1,\n",
    "        cmap='fall'\n",
    "    )\n",
    "\n",
    "    fig = plt.figure(figsize=(2.5, 2))\n",
    "    gs00 = GridSpec(nrows = 1, ncols = 2, width_ratios = [20] + [1], wspace=0.1)\n",
    "    gs_cb = GridSpecFromSubplotSpec(nrows=1, ncols=1, wspace = 0.1, subplot_spec = gs00[1])\n",
    "\n",
    "    #how many heatmaps\n",
    "    gs0 = GridSpecFromSubplotSpec(nrows=1, ncols=1, wspace = 0.3, subplot_spec = gs00[0])\n",
    "    gs1 = GridSpecFromSubplotSpec(nrows = 1, ncols = 1, wspace=0.1, hspace = 0.05, subplot_spec = gs0[0])\n",
    "\n",
    "    c = clrs5Mb[cond]\n",
    "    cis = c.matrix(balance = True)[0:, 0:]\n",
    "    \n",
    "    n,j = np.indices(cis.shape)\n",
    "    n = n.flatten()\n",
    "    j = j.flatten()\n",
    "    # fill NaNs inside of the desired selection:\n",
    "    nan_band_selector = (np.abs(n-j) < 2) #fills 2 bins on diagonal with nan\n",
    "    cis[n[nan_band_selector],j[nan_band_selector]] = np.nan\n",
    "    \n",
    "    cisbins = c.bins()[0:]\n",
    "    \n",
    "    #plot heatmap\n",
    "    ax = plt.subplot(gs1[0, 0])\n",
    "    img = ax.matshow( \n",
    "        np.log10(cis),\n",
    "        **opts)\n",
    "    ax.axes.xaxis.set_visible(False)   \n",
    "    ax.axes.yaxis.set_visible(False)\n",
    "\n",
    "    ax.set_aspect('equal')\n",
    "    plt.title(samplePlotNames[cond] + '\\n')\n",
    "    \n",
    "    #color bar in it's own axis, and own gridspace\n",
    "    colorAx = plt.subplot(gs_cb[0, 0])\n",
    "    cb = plt.colorbar(img, cax = colorAx, extend = 'both')\n",
    "    cb.set_label('log10(interactions)')\n",
    "    #save separately for each condition\n",
    "    plt.savefig(f\"{outDataDir}/figures/{cond}_R1R2_AllbyAllHeatmaps_500MbBins_MinusDiag.png\", dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Difference heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"ticks\")\n",
    "sns.set_context(\"paper\")\n",
    "\n",
    "#ratios for main figure - R1 + R2 combined only, 5Mb bins\n",
    "binsize = 5000000\n",
    "\n",
    "for DMSOcond, Treatcond in zip(ComboCtrlConds, ComboTreatConds):\n",
    "    \n",
    "    opts = dict(\n",
    "       vmin=-0.4,\n",
    "       vmax=0.4,\n",
    "       cmap='coolwarm'\n",
    "    )\n",
    "\n",
    "    fig = plt.figure(figsize=(2.5, 2))\n",
    "\n",
    "    gs00 = GridSpec(nrows = 1, ncols = 2, width_ratios = [20] + [1], wspace=0.1)\n",
    "\n",
    "    gs_cb = GridSpecFromSubplotSpec(nrows=1, ncols=1, wspace = 0.1, subplot_spec = gs00[1])\n",
    "\n",
    "    #how many heatmaps\n",
    "    gs0 = GridSpecFromSubplotSpec(nrows=1, ncols=1, wspace = 0.3, subplot_spec = gs00[0])\n",
    "    \n",
    "    treatc = clrs5Mb[Treatcond]\n",
    "    treatcis = treatc.matrix(balance = True)[0:, 0:]\n",
    "    treatcisbins = treatc.bins()[0:]\n",
    "    \n",
    "    treatLog10 = np.log10(treatcis)\n",
    "    \n",
    "    dmsoc = clrs5Mb[DMSOcond]\n",
    "    dmsocis = dmsoc.matrix(balance = True)[0:, 0:]\n",
    "    dmsocisbins = dmsoc.bins()[0:]\n",
    "    \n",
    "    dmsoLog10 = np.log10(dmsocis)\n",
    "    \n",
    "    ratiocis = treatLog10-dmsoLog10\n",
    "    k,j = np.indices(ratiocis.shape)\n",
    "    k = k.flatten()\n",
    "    j = j.flatten()\n",
    "    # fill NaNs inside of the desired selection:\n",
    "    nan_band_selector = (np.abs(k-j) < 2) #fills 2 bins on diagonal with nan\n",
    "    ratiocis[k[nan_band_selector],j[nan_band_selector]] = np.nan\n",
    "    \n",
    "    ax = plt.subplot(gs0[0])\n",
    "    img = ax.matshow(\n",
    "        ratiocis, \n",
    "        **opts)\n",
    "    ax.axes.xaxis.set_visible(False)   \n",
    "    ax.axes.yaxis.set_visible(False)\n",
    "    plt.title(samplePlotNames[Treatcond] + '/\\n' + samplePlotNames[DMSOcond])\n",
    "    ax.set_aspect('equal')\n",
    "    \n",
    "    # color bar in it's own axis, and own gridspace\n",
    "    colorAx = plt.subplot(gs00[1])\n",
    "    cb = plt.colorbar(img, cax = colorAx, extend = 'both')\n",
    "    cb.set_label('log10(Treated/Ctrl)')\n",
    "\n",
    "    plt.savefig(f\"{outDataDir}/figures/{Treatcond}vs{DMSOcond}_R1R2_AllbyAll_minusdiag_Ratio_500KbBins.png\", dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#chrom 1-5 only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"ticks\")\n",
    "sns.set_context(\"paper\")\n",
    "\n",
    "binsize = 5000000\n",
    "\n",
    "for cond in ComboConds:\n",
    "    \n",
    "    opts = dict(\n",
    "        vmin=-3.5, #change scale range here\n",
    "        vmax=-1,\n",
    "        cmap='fall'\n",
    "    )\n",
    "\n",
    "    fig = plt.figure(figsize=(2.5, 2))\n",
    "\n",
    "    gs00 = GridSpec(nrows = 1, ncols = 2, width_ratios = [20] + [1], wspace=0.1)\n",
    "\n",
    "    gs_cb = GridSpecFromSubplotSpec(nrows=1, ncols=1, wspace = 0.1, subplot_spec = gs00[1])\n",
    "\n",
    "    #how many heatmaps\n",
    "    gs0 = GridSpecFromSubplotSpec(nrows=1, ncols=1, wspace = 0.3, subplot_spec = gs00[0])\n",
    "\n",
    "    gs1 = GridSpecFromSubplotSpec(nrows = 1, ncols = 1, wspace=0.1, hspace = 0.05, subplot_spec = gs0[0])\n",
    "\n",
    "    c = clrs5Mb[cond]\n",
    "    cis = c.matrix(balance = True)[0:214, 0:214]\n",
    "    \n",
    "    n,j = np.indices(cis.shape)\n",
    "    n = n.flatten()\n",
    "    j = j.flatten()\n",
    "    # fill NaNs inside of the desired selection:\n",
    "    nan_band_selector = (np.abs(n-j) < 2) #fills 2 bins on diagonal with nan\n",
    "    cis[n[nan_band_selector],j[nan_band_selector]] = np.nan\n",
    "    \n",
    "    cisbins = c.bins()[0:214]\n",
    "    \n",
    "    #plot heatmap\n",
    "    ax = plt.subplot(gs1[0, 0])\n",
    "    img = ax.matshow( #can change this to plt.imshow instead probably?\n",
    "        np.log10(cis),\n",
    "        **opts)\n",
    "    ax.axes.xaxis.set_visible(False)   \n",
    "    ax.axes.yaxis.set_visible(False)\n",
    "\n",
    "    ax.set_aspect('equal')\n",
    "    plt.title(samplePlotNames[cond] + '\\n')\n",
    "    \n",
    "# color bar in it's own axis, and own gridspace\n",
    "    colorAx = plt.subplot(gs_cb[0, 0])\n",
    "    cb = plt.colorbar(img, cax = colorAx, extend = 'both')\n",
    "    cb.set_label('log10(interactions)')\n",
    "\n",
    "    #plt.suptitle(f'Chr14 Heatmaps and Eigen1 200kb Bins')\n",
    "    plt.savefig(f\"{outDataDir}/figures/{cond}_R1R2_chr1to5_heatmaps_500MbBins_MinusDiag.png\", dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Difference heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"ticks\")\n",
    "sns.set_context(\"paper\")\n",
    "\n",
    "#ratios for main figure - R1 + R2 combined only, 5Mb bins\n",
    "binsize = 5000000\n",
    "\n",
    "for DMSOcond, Treatcond in zip(ComboCtrlConds, ComboTreatConds):\n",
    "    \n",
    "    opts = dict(\n",
    "       vmin=-0.4,\n",
    "       vmax=0.4,\n",
    "        cmap='coolwarm'\n",
    "    )\n",
    "\n",
    "    fig = plt.figure(figsize=(2.5, 2))\n",
    "\n",
    "    gs00 = GridSpec(nrows = 1, ncols = 2, width_ratios = [20] + [1], wspace=0.1)\n",
    "\n",
    "    gs_cb = GridSpecFromSubplotSpec(nrows=1, ncols=1, wspace = 0.1, subplot_spec = gs00[1])\n",
    "\n",
    "    #how many heatmaps\n",
    "    gs0 = GridSpecFromSubplotSpec(nrows=1, ncols=1, wspace = 0.3, subplot_spec = gs00[0])\n",
    "\n",
    "    \n",
    "    treatc = clrs5Mb[Treatcond]\n",
    "    treatcis = treatc.matrix(balance = True)[0:214, 0:214]\n",
    "    treatcisbins = treatc.bins()[0:214]\n",
    "    \n",
    "    treatLog10 = np.log10(treatcis)\n",
    "    \n",
    "    dmsoc = clrs5Mb[DMSOcond]\n",
    "    dmsocis = dmsoc.matrix(balance = True)[0:214, 0:214]\n",
    "    dmsocisbins = dmsoc.bins()[0:214]\n",
    "    \n",
    "    dmsoLog10 = np.log10(dmsocis)\n",
    "    \n",
    "    ratiocis = treatLog10-dmsoLog10\n",
    "    k,j = np.indices(ratiocis.shape)\n",
    "    k = k.flatten()\n",
    "    j = j.flatten()\n",
    "    # fill NaNs inside of the desired selection:\n",
    "    nan_band_selector = (np.abs(k-j) < 2) #fills 2 bins on diagonal with nan\n",
    "    ratiocis[k[nan_band_selector],j[nan_band_selector]] = np.nan\n",
    "    \n",
    "    ax = plt.subplot(gs0[0])\n",
    "    img = ax.matshow(\n",
    "        ratiocis, \n",
    "        **opts)\n",
    "    ax.axes.xaxis.set_visible(False)   \n",
    "    ax.axes.yaxis.set_visible(False)\n",
    "    plt.title(samplePlotNames[Treatcond] + '/\\n' + samplePlotNames[DMSOcond])\n",
    "    ax.set_aspect('equal')\n",
    "    \n",
    "    # color bar in it's own axis, and own gridspace\n",
    "    colorAx = plt.subplot(gs00[1])\n",
    "    cb = plt.colorbar(img, cax = colorAx, extend = 'both')\n",
    "    cb.set_label('log10(Treated/Ctrl)')\n",
    "\n",
    "    plt.savefig(f\"{outDataDir}/figures/{Treatcond}vs{DMSOcond}_R1R2_chr1to5_minusdiag_Ratio_500KbBins.png\", dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Heatmaps with eigenvectors - 250kb bins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#coolers - 250kb bins\n",
    "binsize = 250000\n",
    "\n",
    "clr_paths_250kb = {}\n",
    "for cond in conditions:\n",
    "    clr_paths_250kb[cond] = f'{outDataDir}/data/{long_names[cond]}.sampled.hg38.mapq_30.1000.mcool::resolutions/{binsize}'\n",
    "    \n",
    "clrs250kb = {\n",
    "    cond: cooler.Cooler(clr_paths_250kb[cond]) for cond in conditions\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use bioframe to fetch the genomic features from the UCSC.\n",
    "hg38_chromsizes = bioframe.fetch_chromsizes('hg38')\n",
    "hg38_cens = bioframe.fetch_centromeres('hg38')\n",
    "# create a view with chromosome arms using chromosome sizes and definition of centromeres\n",
    "hg38_arms = bioframe.make_chromarms(hg38_chromsizes,  hg38_cens)\n",
    "hg38_fullchroms = bioframe.make_viewframe(hg38_chromsizes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "good_chroms = ['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr9', 'chr10', 'chr11', 'chr12', 'chr13',\n",
    "              'chr14', 'chr16', 'chr18', 'chr19', 'chr20', 'chr21']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select only chromosomes that are present in the good chromosomes\n",
    "hg38_chromsizes = hg38_chromsizes.loc[good_chroms]\n",
    "hg38_arms = hg38_arms[hg38_arms.chrom.isin(good_chroms)].reset_index(drop=True)\n",
    "hg38_fullchroms = hg38_fullchroms[hg38_fullchroms.chrom.isin(good_chroms)].reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = cooler.binnify(hg38_chromsizes, binsize)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from bioframe.io.resources import UCSCClient\n",
    "mrna = UCSCClient('hg38').fetch_mrna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genecov = bioframe.frac_gene_coverage(bins, mrna) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get compartments, flip based on gene density, all chromosomes\n",
    "#Not sorting by variance compared to gene density, instead by variance explained of data\n",
    "\n",
    "lam = {}\n",
    "eigs = {}\n",
    "\n",
    "for cond in conditions:\n",
    "    lam[cond], eigs[cond] = cooltools.eigs_cis(\n",
    "        clrs250kb[cond], \n",
    "        genecov,\n",
    "        n_eigs=3, \n",
    "        ignore_diags=2,\n",
    "        view_df = hg38_fullchroms\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for cond in conditions:\n",
    "    # Save text files\n",
    "    lam[cond].to_csv(f'{outDataDir}/data/{long_names[cond]}.{binsize//1000}kb.mapq30.byarm.eigs.cis.lam.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot eig1 for each region to make sure they were detecting compartments and not something else...\n",
    "fig = plt.figure(figsize=(15, 2 * len(hg38_fullchroms)))\n",
    "\n",
    "gs1 = GridSpec(nrows = len(hg38_fullchroms), ncols = 1, hspace = 0.8)\n",
    "\n",
    "for i, chrom in enumerate(hg38_fullchroms.itertuples()):\n",
    "    ax = plt.subplot(gs1[i])\n",
    "    for cond in conditions:\n",
    "        loc_eig = bioframe.select(eigs[cond], chrom[1:4])\n",
    "        ax.plot(\n",
    "            loc_eig['start'],\n",
    "            loc_eig['E1'],\n",
    "            label = cond,\n",
    "            color = sampleColors[cond],\n",
    "            ls = sampleLineStyles[cond]\n",
    "        )\n",
    "        plt.axhline(0,ls='--',lw=0.5,color='gray')\n",
    "        plt.ylabel('E1')\n",
    "        plt.xlabel('position, bp')\n",
    "        plt.title(chrom)\n",
    "        plt.legend()\n",
    "        plt.ylim(-2, 2)\n",
    "\n",
    "plt.savefig(f'{outDataDir}/figures/Eig1_250kb_bychrom_mapq30.png', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Looks good except for chr18 and chr21 - need to flip Top2B KO for chr18 (all) and WT Ctrl R1 + R2, need to flip Top2BKO ATRA R1 and R1R2 for chr21.\n",
    "\n",
    "for cond in ['WT_ATRA_R1', 'WT_ATRA_R2', 'WT_ATRA_R1R2', ]:\n",
    "    eigs[cond].loc[eigs[cond]['chrom'] == 'chr18', 'E1'] = (eigs[cond].loc[eigs[cond]['chrom'] == 'chr18', 'E1'])*-1\n",
    "    \n",
    "for cond in ['Top2BKO_ATRA_R1', 'Top2BKO_ATRA_R1R2']:\n",
    "    eigs[cond].loc[eigs[cond]['chrom'] == 'chr21', 'E1'] = (eigs[cond].loc[eigs[cond]['chrom'] == 'chr21', 'E1'])*-1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot eig1 for each region to make sure they were detecting compartments and not something else...\n",
    "fig = plt.figure(figsize=(15, 2 * len(hg38_fullchroms)))\n",
    "\n",
    "gs1 = GridSpec(nrows = len(hg38_fullchroms), ncols = 1, hspace = 0.8)\n",
    "\n",
    "for i, chrom in enumerate(hg38_fullchroms.itertuples()):\n",
    "    ax = plt.subplot(gs1[i])\n",
    "    for cond in conditions:\n",
    "        loc_eig = bioframe.select(eigs[cond], chrom[1:4])\n",
    "        ax.plot(\n",
    "            loc_eig['start'],\n",
    "            loc_eig['E1'],\n",
    "            label = cond,\n",
    "            color = sampleColors[cond],\n",
    "            ls = sampleLineStyles[cond]\n",
    "        )\n",
    "        plt.axhline(0,ls='--',lw=0.5,color='gray')\n",
    "        plt.ylabel('E1')\n",
    "        plt.xlabel('position, bp')\n",
    "        plt.title(chrom)\n",
    "        plt.legend()\n",
    "        plt.ylim(-2, 2)\n",
    "\n",
    "plt.savefig(f'{outDataDir}/figures/Corrected_Eig1_250kb_bychrom_mapq30.png', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for cond in conditions:\n",
    "    # Save text files\n",
    "    eigs[cond].to_csv(f'{outDataDir}/data/{long_names[cond]}.{binsize//1000}kb.mapq30.byarm.eigs.cis.vecs.txt', sep='\\t', index=False, na_rep = 'nan')\n",
    "    \n",
    "    # Save bedGraph track\n",
    "    eigs[cond][['chrom', 'start', 'end', 'E1']].to_csv(f'{outDataDir}/data/{long_names[cond]}.{binsize//1000}kb.mapq30.byarm.eigs.cis.vecs.E1.bedGraph', sep='\\t', index=False, na_rep = 'nan', header = False)\n",
    "    # Save bigwig track\n",
    "    bioframe.to_bigwig(eigs[cond], hg38_chromsizes, f'{outDataDir}/data/{long_names[cond]}.{binsize//1000}kb.mapq30.byarm.eigs.cis.vecs.E1.bw', 'E1')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in eigs\n",
    "eigs = {}\n",
    "for cond in conditions:\n",
    "    eigs[cond] = pd.read_csv(f'{outDataDir}/data/{long_names[cond]}.{binsize//1000}kb.mapq30.byarm.eigs.cis.vecs.txt', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cooltools.lib.plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_context(\"paper\")\n",
    "sns.set_style('ticks')\n",
    "\n",
    "for cond in ComboConds:\n",
    "    \n",
    "    plottingregion = 'chr14:17086761-107043718'\n",
    "    chromarm = ('chr14', 17086761, 107043718)\n",
    "\n",
    "    opts = dict(\n",
    "        vmin=-4, #change scale range here\n",
    "        vmax=-2,\n",
    "        cmap='fall'\n",
    "    )\n",
    "\n",
    "    fig = plt.figure(figsize=(2.5, 2.5))\n",
    "\n",
    "    gs00 = GridSpec(nrows = 1, ncols = 2, width_ratios = [20] + [1], wspace=0.5)\n",
    "\n",
    "    gs_cb = GridSpecFromSubplotSpec(nrows=2, ncols=1, wspace = 0.1, subplot_spec = gs00[1], hspace = 0.05, \n",
    "                                    height_ratios = [6, 1])\n",
    "\n",
    "    #how many heatmaps\n",
    "    gs0 = GridSpecFromSubplotSpec(nrows=1, ncols=1, subplot_spec = gs00[0])\n",
    "    \n",
    "    gs1 = GridSpecFromSubplotSpec(nrows = 2, ncols = 1, height_ratios=[6, 1], wspace=0.1, \n",
    "                                  hspace = 0.05, subplot_spec = gs0[0])\n",
    "\n",
    "    c = clrs250kb[cond]\n",
    "    cis = c.matrix(balance = True).fetch(plottingregion)\n",
    "    \n",
    "    n,j = np.indices(cis.shape)\n",
    "    n = n.flatten()\n",
    "    j = j.flatten()\n",
    "    # fill NaNs inside of the desired selection:\n",
    "    nan_band_selector = (np.abs(n-j) < 2) #fills 2 bins on diagonal with nan\n",
    "    cis[n[nan_band_selector],j[nan_band_selector]] = np.nan\n",
    "    \n",
    "    cisbins = c.bins().fetch(plottingregion)\n",
    "    mids = cisbins['end'] - binsize/2\n",
    "    mids = mids.reset_index(drop=True)\n",
    "    \n",
    "    #plot heatmap\n",
    "    ax = plt.subplot(gs1[0, 0])\n",
    "    img = ax.matshow( #can change this to plt.imshow instead probably?\n",
    "        np.log10(cis), \n",
    "        extent=[\n",
    "            mids[0]/1000000, \n",
    "            mids.iloc[-1]/1000000, \n",
    "            mids.iloc[-1]/1000000, \n",
    "            mids[0]/1000000], \n",
    "        **opts)\n",
    "    ax.xaxis.set_visible(False)\n",
    "    ax.yaxis.tick_right()    \n",
    "    ax.set_aspect('equal')\n",
    "    plt.title(samplePlotNames[cond] + '\\n')\n",
    "    \n",
    "    # barplot of eigenvalues\n",
    "    ax1 = plt.subplot(gs1[1, 0])\n",
    "    plt.ylim(-2, 2) \n",
    "    plt.xlim(chromarm[1]/1000000, chromarm[2]/1000000)\n",
    "    img2 = ax1.bar(\n",
    "        x = list((eigs[cond][eigs[cond].chrom == chromarm[0]]['start'] + binsize/2)/1000000), \n",
    "        height = list(eigs[cond][eigs[cond].chrom == chromarm[0]]['E1']),\n",
    "        width = 0.25,\n",
    "        color=eigs[cond].E1[eigs[cond].chrom == chromarm[0]].apply(lambda x: 'r' if x>0 else 'b'),\n",
    "        alpha = 1,\n",
    "        linewidth = 0\n",
    "                  )\n",
    "    plt.ylabel('Eig 1')\n",
    "    plt.xlabel('Mb')\n",
    "    ax1.xaxis.set_ticks(np.arange(20, 110, 10)) #This specifically sets ticks to this range, to match y axis\n",
    "    fig.add_subplot(ax1)\n",
    "    \n",
    "    # color bar in it's own axis, and own gridspace\n",
    "    colorAx = plt.subplot(gs_cb[0, 0])\n",
    "    cb = plt.colorbar(img, cax = colorAx, extend = 'both')\n",
    "    cb.set_label('log10(interactions)')\n",
    "\n",
    "    plt.savefig(f\"{outDataDir}/figures/{cond}_R1R2Combined_FullChr14Heatmaps_250kbBins_MinusDiag_WithEigen1_Barplot_columns.png\", dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Difference heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ratios for main figure - R1 + R2 combined only, 250kb bins\n",
    "\n",
    "for DMSOcond, Treatcond in zip(ComboCtrlConds, ComboTreatConds):\n",
    "    \n",
    "    plottingregion = 'chr14:17086761-107043718'\n",
    "    chromarm = ('chr14', 17086761, 107043718)\n",
    "\n",
    "    opts = dict(\n",
    "        vmin=-.5,\n",
    "        vmax=.5,\n",
    "        cmap='coolwarm'\n",
    "    )\n",
    "\n",
    "    fig = plt.figure(figsize=(2.5, 2.5))\n",
    "\n",
    "    gs00 = GridSpec(nrows = 1, ncols = 2, width_ratios = [20] + [1], wspace=0.5)\n",
    "\n",
    "    gs_cb = GridSpecFromSubplotSpec(nrows=2, ncols=1, wspace = 0.1, subplot_spec = gs00[1], hspace = 0.05, \n",
    "                                    height_ratios = [6, 1])\n",
    "\n",
    "    #how many heatmaps\n",
    "    gs0 = GridSpecFromSubplotSpec(nrows=1, ncols=1, subplot_spec = gs00[0])\n",
    "    gs1 = GridSpecFromSubplotSpec(nrows = 2, ncols = 1, height_ratios=[6, 1], wspace=0.1, \n",
    "                                  hspace = 0.05, subplot_spec = gs0[0])\n",
    "    \n",
    "    treatc = clrs250kb[Treatcond]\n",
    "    treatcis = treatc.matrix(balance = True).fetch(plottingregion)\n",
    "    treatcisbins = treatc.bins().fetch(plottingregion)\n",
    "    treatmids = treatcisbins['end'] - binsize/2\n",
    "    treatmids = treatmids.reset_index(drop=True)\n",
    "    treatLog10 = np.log10(treatcis)\n",
    "    \n",
    "    dmsoc = clrs250kb[DMSOcond]\n",
    "    dmsocis = dmsoc.matrix(balance = True).fetch(plottingregion)\n",
    "    dmsocisbins = dmsoc.bins().fetch(plottingregion)\n",
    "    dmsomids = dmsocisbins['end'] - binsize/2\n",
    "    dmsomids = dmsomids.reset_index(drop=True)\n",
    "    dmsoLog10 = np.log10(dmsocis)\n",
    "    \n",
    "    ratiocis = treatLog10-dmsoLog10\n",
    "    k,j = np.indices(ratiocis.shape)\n",
    "    k = k.flatten()\n",
    "    j = j.flatten()\n",
    "    # fill NaNs inside of the desired selection:\n",
    "    nan_band_selector = (np.abs(k-j) < 2) #fills 2 bins on diagonal with nan\n",
    "    ratiocis[k[nan_band_selector],j[nan_band_selector]] = np.nan\n",
    "    \n",
    "    ax = plt.subplot(gs1[0, 0])\n",
    "    img = ax.matshow(\n",
    "        ratiocis, extent=[treatmids[0]/1000000, treatmids.iloc[-1]/1000000, treatmids.iloc[-1]/1000000, treatmids[0]/1000000], \n",
    "        **opts)\n",
    "    ax.axes.xaxis.set_visible(False)   \n",
    "    ax.yaxis.tick_right() \n",
    "    plt.title(samplePlotNames[Treatcond] + '/\\n' + samplePlotNames[DMSOcond])\n",
    "    ax.set_aspect('equal')\n",
    "    \n",
    "    # barplot of eigenvalues - control and treatment - lineplots\n",
    "    ax1 = plt.subplot(gs1[1, 0])\n",
    "    plt.ylim(-2, 2) \n",
    "    plt.xlim(chromarm[1]/1000000, chromarm[2]/1000000)\n",
    "    ax1.plot(\n",
    "        list((eigs[DMSOcond][eigs[DMSOcond].chrom == chromarm[0]]['start'] + binsize/2)/1000000), \n",
    "        list(eigs[DMSOcond][eigs[DMSOcond].chrom == chromarm[0]]['E1']), color = sampleColors[DMSOcond], \n",
    "        label = samplePlotNames[DMSOcond])\n",
    "    ax1.plot(\n",
    "        list((eigs[Treatcond][eigs[Treatcond].chrom == chromarm[0]]['start'] + binsize/2)/1000000), \n",
    "        list(eigs[Treatcond][eigs[Treatcond].chrom == chromarm[0]]['E1']), color = sampleColors[Treatcond], \n",
    "        label = samplePlotNames[Treatcond])\n",
    "    plt.ylabel('Eig 1')\n",
    "    plt.xlabel('Mb')\n",
    "    ax1.xaxis.set_ticks(np.arange(20, 110, 10)) #This specifically sets ticks to this range, to match y axis\n",
    "    ax1.legend(bbox_to_anchor=(1.04,1), frameon = False)\n",
    "    fig.add_subplot(ax1)\n",
    "    \n",
    "    # color bar in it's own axis, and own gridspace\n",
    "    colorAx = plt.subplot(gs_cb[0])\n",
    "    cb = plt.colorbar(img, cax = colorAx, extend = 'both')\n",
    "    cb.set_label('log10(Treated/Ctrl)')\n",
    "\n",
    "    plt.savefig(f\"{outDataDir}/figures/{Treatcond}vs{DMSOcond}_R1R2Combined_Chr14DifferenceHeatmaps_minusdiag_250KbBins_LargerColorScale.png\".format(outDataDir), dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Saddleplots and strength with OWN eig1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#coolers - 250kb bins\n",
    "binsize = 250000\n",
    "\n",
    "clr_paths_250kb = {}\n",
    "for cond in conditions:\n",
    "    clr_paths_250kb[cond] = f'{outDataDir}/data/{long_names[cond]}.sampled.hg38.mapq_30.1000.mcool::resolutions/{binsize}'\n",
    "    \n",
    "clrs250kb = {\n",
    "    cond: cooler.Cooler(clr_paths_250kb[cond]) for cond in conditions\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#need to filter out arms without any data - from gfudenberg (https://github.com/open2c/cooltools/issues/330)\n",
    "bins = clrs250kb['WT_Ctrl_R1'].bins()[:].copy()\n",
    "bins['coverage'] = cooltools.coverage(clrs250kb['WT_Ctrl_R1'])[1]\n",
    "bins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "view_coverage = bioframe.overlap(hg38_arms, bins).groupby('name').sum('coverage_')\n",
    "view_coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "view_coverage['coverage_'].values > 1e4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hg38_arms_filtered = bioframe.make_viewframe(hg38_arms[hg38_arms['name'].isin(view_coverage[view_coverage['coverage_'].values > 1e4].index)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hg38_arms_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hg38_arms_filtered.to_csv(f'{outDataDir}/data/hg38_arms_filtered.bed', index = False, header = None, sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for cond in conditions:\n",
    "    in_fname = clr_paths_250kb[cond]\n",
    "    region_fname = f'{outDataDir}/data/hg38_arms_filtered.bed'\n",
    "    out_fname = f'{outDataDir}/data/{long_names[cond]}.{binsize//1000}kb.mapq30.expected.cis.cli.tsv'\n",
    "    !bsub -q short -W 01:00 -e ./lsf_jobs/LSB_%J.err -o ./lsf_jobs/LSB_%J.log \\\n",
    "        -n 4 -R span[hosts=1] -R select[ib] -R rusage[mem=4000] -R select[rh=8] -N \\\n",
    "        \"cooltools expected-cis -p 4 -o $out_fname --clr-weight-name weight --ignore-diags 2 $in_fname --regions $region_fname\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Saddle strength - AA, BB, AA + BB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#AA vs BB compartment scores\n",
    "def saddle_strengthAB(S, C):\n",
    "    \"\"\"\n",
    "    Parameters\n",
    "    ----------\n",
    "    S, C : 2D arrays, square, same shape\n",
    "        Saddle sums and counts, respectively\n",
    "        \n",
    "    Returns\n",
    "    -------\n",
    "    2x1D array\n",
    "    Ratios of cumulative corner interaction scores, where the saddle data is \n",
    "    separately AA/AB and BB/BA corners with increasing extent\n",
    "    \n",
    "    \"\"\"\n",
    "    m, n = S.shape\n",
    "    if m != n:\n",
    "        raise ValueError(\"`saddledata` should be square.\")\n",
    "\n",
    "    ratiosA = np.zeros(n)\n",
    "    for k in range(1, n):\n",
    "        intra_sumA = S[n-k:n, n-k:n].sum() \n",
    "        intra_countA = C[n-k:n, n-k:n].sum()\n",
    "        intraA = intra_sumA / intra_countA\n",
    "        \n",
    "        inter_sum = S[0:k, n-k:n].sum() + S[n-k:n, 0:k].sum()\n",
    "        inter_count =  C[0:k, n-k:n].sum() + C[n-k:n, 0:k].sum()\n",
    "        inter = inter_sum / inter_count\n",
    "        \n",
    "        ratiosA[k] = intraA / inter\n",
    "        \n",
    "    ratiosB = np.zeros(n)\n",
    "    for k in range(1, n):\n",
    "        intra_sumB = S[0:k, 0:k].sum()\n",
    "        intra_countB = C[0:k, 0:k].sum()\n",
    "        intraB = intra_sumB / intra_countB\n",
    "        \n",
    "        inter_sum = S[0:k, n-k:n].sum() + S[n-k:n, 0:k].sum()\n",
    "        inter_count =  C[0:k, n-k:n].sum() + C[n-k:n, 0:k].sum()\n",
    "        inter = inter_sum / inter_count\n",
    "        \n",
    "        ratiosB[k] = intraB / inter\n",
    "    \n",
    "    ratios = {\n",
    "        'A' : ratiosA,\n",
    "        'B' : ratiosB\n",
    "    }\n",
    "    \n",
    "    return ratios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use this if expected already run, will be much faster\n",
    "cis_exp = {}\n",
    "\n",
    "for cond in conditions:\n",
    "    cis_exp[cond] = pd.read_csv(f'{outDataDir}/data/{long_names[cond]}.{binsize//1000}kb.mapq30.expected.cis.cli.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cis_exp[cond] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eigs = {}\n",
    "for cond in conditions:\n",
    "    eigs[cond] = pd.read_csv(f'{outDataDir}/data/{long_names[cond]}.{binsize//1000}kb.mapq30.byarm.eigs.cis.vecs.txt', sep = '\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eigs[cond]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Saddle strength - all distances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Use the function to return saddledata (just good chromosomes, by arm)\n",
    "\n",
    "sums_ownE1 = {}\n",
    "counts_ownE1 = {}\n",
    "\n",
    "for cond in conditions:\n",
    "    sums_ownE1[cond], counts_ownE1[cond] = cooltools.api.saddle.saddle(\n",
    "            clr = clrs250kb[cond], \n",
    "            expected = cis_exp[cond], \n",
    "            contact_type = 'cis',\n",
    "            view_df = hg38_arms_filtered.reset_index(drop = True),\n",
    "            track = eigs[cond][['chrom', 'start', 'end', 'E1']],\n",
    "            qrange = (0.02, 0.98),\n",
    "            n_bins = 50,\n",
    "            verbose = False\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Compartment strength - AA vs BB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Own Eig1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#AA vs BB strength\n",
    "strengthAB_ownE1 = {}\n",
    "for cond in conditions:\n",
    "    strengthAB_ownE1[cond] = saddle_strengthAB(sums_ownE1[cond], counts_ownE1[cond])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SepTreatConds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SepCtrlConds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#log2 ratio of treatment/control saddle strength within each replicate\n",
    "sep_log2_treatvsctrl = {}\n",
    "for comp in ['A', 'B']:\n",
    "    sep_log2_treatvsctrl[comp] = {}\n",
    "    for treat, ctrl in zip(SepTreatConds, SepCtrlConds):\n",
    "        sep_log2_treatvsctrl[comp][f'{treat}vs{ctrl}'] = np.log2(strengthAB_ownE1[treat][comp]) - np.log2(strengthAB_ownE1[ctrl][comp])\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot bargraph of avg and dots of replicates log2(treat/ctrl) for AA and BB comp strength\n",
    "#10 bin square - top 20% A or B\n",
    "\n",
    "comp_score_df = pd.DataFrame(columns = ['Comparison', 'Compartment', 'Replicate', 'Label', 'Score'])\n",
    "\n",
    "repdict = {\n",
    "    'WT_Ctrl_R1' : 'R1',\n",
    "    'Top2BKO_Ctrl_R1' : 'R1',\n",
    "    'WT_ATRA_R1' : 'R1',\n",
    "    'Top2BKO_ATRA_R1' : 'R1',    \n",
    "    'WT_Ctrl_R2' : 'R2',\n",
    "    'Top2BKO_Ctrl_R2' : 'R2',\n",
    "    'WT_ATRA_R2' : 'R2',\n",
    "    'Top2BKO_ATRA_R2' : 'R2',\n",
    "}\n",
    "\n",
    "labeldict = {\n",
    "    'WT_Ctrl_R1' : 'WT Ctrl',\n",
    "    'Top2BKO_Ctrl_R1' : 'BKO Ctrl',\n",
    "    'WT_ATRA_R1' : 'WT ATRA',\n",
    "    'Top2BKO_ATRA_R1' : 'BKO ATRA',\n",
    "    'WT_Ctrl_R2' : 'WT Ctrl',\n",
    "    'Top2BKO_Ctrl_R2' : 'BKO Ctrl',\n",
    "    'WT_ATRA_R2' : 'WT ATRA',\n",
    "    'Top2BKO_ATRA_R2' : 'BKO ATRA',\n",
    "}\n",
    "\n",
    "labelPlotColors = {\n",
    "    'WT Ctrl' : '#17BECF',\n",
    "    'BKO Ctrl' : '#D62728',\n",
    "    'WT ATRA' : '#574D68',\n",
    "    'BKO ATRA' : '#C6A15B',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot actual comp strength (not vs control), cis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "comp_score_df_cis = pd.DataFrame(columns = ['Condition', 'Compartment', 'Replicate', 'Label', 'Score'])\n",
    "\n",
    "for cond in conditions[0:8]:\n",
    "    for comp in ['A', 'B']:\n",
    "            comp_score_df_cis = comp_score_df_cis.append({\n",
    "                'Condition' : cond, \n",
    "                'Compartment' : comp, \n",
    "                'Replicate' : repdict[cond], \n",
    "                'Label' : labeldict[cond],\n",
    "                'Score' : strengthAB_ownE1[cond][comp][9]\n",
    "            }, ignore_index = True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#https://stackoverflow.com/questions/64223870/seaborn-overlap-swarmplot-on-barplot\n",
    "sns.set_style(\"ticks\")\n",
    "sns.set_context(\"paper\")\n",
    "cmap_bar = sns.color_palette(['#a6cee3', '#1f78b4', '#b2df8a', '#33a02c'])\n",
    "gs = GridSpec(nrows= 1, ncols=1, wspace = 0.6, hspace = 0.6)\n",
    "\n",
    "plt.figure(figsize=(2, 2))\n",
    "\n",
    "ax = plt.subplot(gs[0])\n",
    "\n",
    "sns.stripplot(x='Compartment', \n",
    "                  y='Score', \n",
    "                  hue='Label', \n",
    "                  dodge=True, \n",
    "                  data=comp_score_df_cis, \n",
    "                  jitter = False, \n",
    "                  palette = cmap_bar, \n",
    "                  ax = ax)\n",
    "ax1 = sns.barplot(x='Compartment', \n",
    "                      y='Score', \n",
    "                      hue='Label', \n",
    "                      data=comp_score_df_cis, \n",
    "                      palette = cmap_bar, \n",
    "                      alpha = 0.5, \n",
    "                      ci = False, \n",
    "                      ax = ax)\n",
    "    \n",
    "handles, labels = ax1.get_legend_handles_labels()\n",
    "ax1.legend(handles[4:8], labels[4:8], bbox_to_anchor=(1.04,1), frameon = False)\n",
    " \n",
    "plt.title(f'Intrachromosomal\\nCompartment Strength')\n",
    "plt.ylabel('Saddle Strength')  \n",
    "plt.ylim(0, 10)\n",
    "plt.xlabel('Compartment')\n",
    "        \n",
    "plt.savefig(f'{outDataDir}/figures/R1R2_HiC_compstrength_notvsctrl_Eig1.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Compartment strength analysis - use own eigens, calculate separately for reach replicate, trans saddle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create pairwise combinations of chromosomes for calculating average interactions:\n",
    "regions = bioframe.core.construction.add_ucsc_name_column(bioframe.from_any(hg38_chromsizes))\n",
    "region_pairs = list(combinations(regions[['chrom', 'start', 'end']].values, 2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "region_pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "regions.to_csv(f'{outDataDir}/data/hg38_chroms.bed', index = False, sep = '\\t', header = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for cond in conditions:\n",
    "    in_fname = clr_paths_250kb[cond]\n",
    "    region_fname = f'{outDataDir}/data/hg38_chroms.bed'\n",
    "    out_fname = f'{outDataDir}/data/{long_names[cond]}.{binsize//1000}kb.mapq30.expected.trans.cli.tsv'\n",
    "    !bsub -q short -W 01:00 -e ./lsf_jobs/LSB_%J.err -o ./lsf_jobs/LSB_%J.log \\\n",
    "        -n 4 -R span[hosts=1] -R select[ib] -R rusage[mem=4000] -R select[rh=8] -N \\\n",
    "        \"cooltools expected-trans -p 4 -o $out_fname $in_fname --view $region_fname\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Saddle strength - AA, BB, AA + BB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use this if expected already run, will be much faster\n",
    "trans_exp = {}\n",
    "\n",
    "for cond in conditions:\n",
    "    trans_exp[cond] = pd.read_csv(f'{outDataDir}/data/{long_names[cond]}.{binsize//1000}kb.mapq30.expected.trans.cli.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trans_exp[cond]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Use the function to return saddledata for each band size (just good chromosomes, by arm)\n",
    "\n",
    "trans_sums = {}\n",
    "trans_counts = {}\n",
    "\n",
    "for cond in conditions:\n",
    "    trans_sums[cond], trans_counts[cond] = cooltools.api.saddle.saddle(\n",
    "        clr = clrs250kb[cond], \n",
    "        expected = trans_exp[cond], \n",
    "        contact_type = 'trans',\n",
    "        view_df = regions,\n",
    "        track = eigs[cond][['chrom', 'start', 'end', 'E1']],\n",
    "        qrange = (0.02, 0.98),\n",
    "        n_bins = 50,\n",
    "        verbose = False\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Compartment strength - AA vs BB, in trans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#AA vs BB compartment scores\n",
    "def saddle_strengthAB(S, C):\n",
    "    \"\"\"\n",
    "    Parameters\n",
    "    ----------\n",
    "    S, C : 2D arrays, square, same shape\n",
    "        Saddle sums and counts, respectively\n",
    "        \n",
    "    Returns\n",
    "    -------\n",
    "    2x1D array\n",
    "    Ratios of cumulative corner interaction scores, where the saddle data is \n",
    "    separately AA/AB and BB/BA corners with increasing extent\n",
    "    \n",
    "    \"\"\"\n",
    "    m, n = S.shape\n",
    "    if m != n:\n",
    "        raise ValueError(\"`saddledata` should be square.\")\n",
    "\n",
    "    ratiosA = np.zeros(n)\n",
    "    for k in range(1, n):\n",
    "        intra_sumA = S[n-k:n, n-k:n].sum() \n",
    "        intra_countA = C[n-k:n, n-k:n].sum()\n",
    "        intraA = intra_sumA / intra_countA\n",
    "        \n",
    "        inter_sum = S[0:k, n-k:n].sum() + S[n-k:n, 0:k].sum()\n",
    "        inter_count =  C[0:k, n-k:n].sum() + C[n-k:n, 0:k].sum()\n",
    "        inter = inter_sum / inter_count\n",
    "        \n",
    "        ratiosA[k] = intraA / inter\n",
    "        \n",
    "    ratiosB = np.zeros(n)\n",
    "    for k in range(1, n):\n",
    "        intra_sumB = S[0:k, 0:k].sum()\n",
    "        intra_countB = C[0:k, 0:k].sum()\n",
    "        intraB = intra_sumB / intra_countB\n",
    "        \n",
    "        inter_sum = S[0:k, n-k:n].sum() + S[n-k:n, 0:k].sum()\n",
    "        inter_count =  C[0:k, n-k:n].sum() + C[n-k:n, 0:k].sum()\n",
    "        inter = inter_sum / inter_count\n",
    "        \n",
    "        ratiosB[k] = intraB / inter\n",
    "    \n",
    "    ratios = {\n",
    "        'A' : ratiosA,\n",
    "        'B' : ratiosB\n",
    "    }\n",
    "    \n",
    "    return ratios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Own Eig1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#AA vs BB strength, trans\n",
    "strengthABTrans_ownE1 = {}\n",
    "for cond in conditions:\n",
    "    strengthABTrans_ownE1[cond] = saddle_strengthAB(trans_sums[cond], trans_counts[cond])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot bargraph of avg and dots of replicates - not vs control for AA and BB comp strength, trans\n",
    "#5 bin square - top 10% A or B\n",
    "\n",
    "comp_score_df_trans = pd.DataFrame(columns = ['Condition', 'Compartment', 'Replicate', 'Label', 'Score'])\n",
    "\n",
    "for cond in conditions[0:8]:\n",
    "    for comp in ['A', 'B']:\n",
    "        comp_score_df_trans = comp_score_df_trans.append({\n",
    "            'Condition' : cond, \n",
    "            'Compartment' : comp, \n",
    "            'Replicate' : repdict[cond], \n",
    "            'Label' : labeldict[cond],\n",
    "            'Score' : strengthABTrans_ownE1[cond][comp][4]\n",
    "        }, ignore_index = True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#https://stackoverflow.com/questions/64223870/seaborn-overlap-swarmplot-on-barplot\n",
    "sns.set_style(\"ticks\")\n",
    "sns.set_context(\"paper\")\n",
    "cmap_bar = sns.color_palette(['#a6cee3', '#1f78b4', '#b2df8a', '#33a02c'])\n",
    "gs = GridSpec(nrows= 1, ncols=1, wspace = 0.6, hspace = 0.6)\n",
    "\n",
    "plt.figure(figsize=(2, 2))\n",
    "\n",
    "sns.stripplot(x='Compartment', y='Score', hue='Label', dodge=True, data=comp_score_df_trans, jitter = False, palette = cmap_bar)\n",
    "ax1 = sns.barplot(x='Compartment', y='Score', hue='Label', data=comp_score_df_trans, palette = cmap_bar, alpha = 0.5, ci = False)\n",
    "    \n",
    "handles, labels = ax1.get_legend_handles_labels()\n",
    "ax1.legend(handles[4:8], labels[4:8], bbox_to_anchor=(1.04,1), loc=\"upper left\", frameon = False)\n",
    "plt.ylim(0, 6)\n",
    "\n",
    "plt.title(f'Interchromosomal\\nCompartment Strength')\n",
    "plt.ylabel('Saddle Strength')  \n",
    "plt.xlabel('Compartment')\n",
    "        \n",
    "plt.savefig(f'{outDataDir}/figures/HiC_TransCompStrength_AllSamples_BarWithScatter_250kbbinEig1.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Density Plots and Scaling Plots\n",
    "#https://cooltools.readthedocs.io/en/latest/notebooks/contacts_vs_distance.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Calculate expected on 1kb binned coolers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#coolers - 1kb bins\n",
    "binsize = 1000\n",
    "\n",
    "clr_paths_1kb = {}\n",
    "for cond in conditions:\n",
    "    clr_paths_1kb[cond] = f'{outDataDir}/data/{long_names[cond]}.sampled.hg38.mapq_30.1000.mcool::resolutions/{binsize}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for cond in conditions:\n",
    "    in_fname = clr_paths_1kb[cond]\n",
    "    region_fname = f'{outDataDir}/data/hg38_arms_filtered.bed'\n",
    "    out_fname = f'{outDataDir}/data/{long_names[cond]}.{binsize//1000}kb.mapq30.expected.cis.cli.tsv'\n",
    "    !bsub -q short -W 01:00 -e ./lsf_jobs/LSB_%J.err -o ./lsf_jobs/LSB_%J.log \\\n",
    "        -n 8 -R span[hosts=1] -R select[ib] -R rusage[mem=2000] -R select[rh=8] -N \\\n",
    "        \"cooltools expected-cis -p 8 -o $out_fname --ignore-diags 2 --view $region_fname $in_fname\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in log binned expected for plotting\n",
    "exp = {}\n",
    "for cond in conditions:\n",
    "    exp[cond] = pd.read_csv(f'{outDataDir}/data/{long_names[cond]}.{binsize//1000}kb.mapq30.expected.cis.cli.tsv', sep = '\\t')\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "log_exp = {}\n",
    "for cond in conditions:\n",
    "    log_exp[cond] = cooltools.api.expected.logbin_expected(\n",
    "        exp[cond], summary_name='balanced.sum', bins_per_order_magnitude=10, \n",
    "        bin_layout='fixed', min_nvalid=200, min_count=50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "log_exp_agg = {}\n",
    "for cond in conditions:\n",
    "    log_exp_agg[cond] = cooltools.api.expected.combine_binned_expected(\n",
    "        log_exp[cond][0], binned_exp_slope=log_exp[cond][1],\n",
    "        Pc_name='balanced.avg', minmax_drop_bins=2, concat_original=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add Density column - (balanced.avg * n_valid)/sum(balanced.avg * n_valid)\n",
    "for cond in conditions:\n",
    "    log_exp_agg[cond][0]['density'] = (\n",
    "        log_exp_agg[cond][0]['balanced.avg'] * log_exp_agg[cond][0]['n_valid'])/np.nansum(\n",
    "        log_exp_agg[cond][0]['balanced.avg'] * log_exp_agg[cond][0]['n_valid'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "log_exp_agg[cond][0].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Smooth density before plotting, on log bins like for scaling. Then convert back to linear so can use semilogx\n",
    "sns.set_style(\"ticks\")\n",
    "sns.set_context(\"paper\")\n",
    "\n",
    "from cooltools.lib import numutils\n",
    "smoothf = der_smooth_function_combined=lambda x: numutils.robust_gauss_filter(x, 1.3)\n",
    "\n",
    "f, ax = plt.subplots(\n",
    "    figsize=(2, 2))\n",
    "\n",
    "for cond in ComboConds:\n",
    "    ax.semilogx(\n",
    "        10**smoothf(np.log10(log_exp_agg[cond][0]['dist.avg']*1000)),\n",
    "        smoothf(log_exp_agg[cond][0]['density']),\n",
    "        color = sampleColors[cond],\n",
    "        label = samplePlotNames[cond],\n",
    "        ls = sampleLineStyles[cond]\n",
    "    )\n",
    "\n",
    "    ax.set(\n",
    "        xlabel='separation, bp',\n",
    "        ylabel='Density',\n",
    "        xlim=(1e3,1e8),\n",
    "        #ylim=(0, 0.07)\n",
    "    )\n",
    "\n",
    "    ax.grid(lw=0.5)\n",
    "    ax.legend(bbox_to_anchor=(1.04,1), loc=\"upper left\", frameon = False)\n",
    "    #ax.set_aspect(100)\n",
    "    \n",
    "plt.savefig(f'{outDataDir}/figures/R1R2_Density.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot scaling - combo with line at mitotic dropoff\n",
    "sns.set_style(\"ticks\")\n",
    "sns.set_context(\"paper\")\n",
    "\n",
    "f, ax = plt.subplots(\n",
    "    figsize=(2, 2))\n",
    "\n",
    "for cond in ComboConds:\n",
    "    ax.loglog(\n",
    "        log_exp_agg[cond][0]['dist.avg']*1000,\n",
    "        log_exp_agg[cond][0]['balanced.avg'],\n",
    "        color = sampleColors[cond],\n",
    "        label = samplePlotNames[cond],\n",
    "        ls = sampleLineStyles[cond],\n",
    "        lw = 2\n",
    "    )\n",
    "\n",
    "ax.set(\n",
    "    xlabel='separation (s), bp',\n",
    "    ylabel='P(s)',\n",
    "    xlim=(1e3,1e8))\n",
    "\n",
    "ax.grid(lw=0.5)\n",
    "ax.legend(bbox_to_anchor=(1.04,1), loc=\"upper left\", frameon = False)\n",
    "plt.title('P(s) Scaling Plot')\n",
    "\n",
    "plt.axvline(x=1.5e7)\n",
    "    \n",
    "plt.savefig(f'{outDataDir}/figures/R1R2_Scaling_Combo_With1.5e7vline.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot scaling - combo \n",
    "sns.set_style(\"ticks\")\n",
    "sns.set_context(\"paper\")\n",
    "\n",
    "f, ax = plt.subplots(\n",
    "    figsize=(2, 2))\n",
    "\n",
    "for cond in ComboConds:\n",
    "    ax.loglog(\n",
    "        log_exp_agg[cond][0]['dist.avg']*1000,\n",
    "        log_exp_agg[cond][0]['balanced.avg'],\n",
    "        color = sampleColors[cond],\n",
    "        label = samplePlotNames[cond],\n",
    "        ls = sampleLineStyles[cond],\n",
    "        lw = 2\n",
    "    )\n",
    "\n",
    "ax.set(\n",
    "    xlabel='separation (s), bp',\n",
    "    ylabel='P(s)',\n",
    "    xlim=(1e3,1e8))\n",
    "\n",
    "ax.grid(lw=0.5)\n",
    "ax.legend(bbox_to_anchor=(1.04,1), loc=\"upper left\", frameon = False)\n",
    "plt.title('P(s) Scaling Plot')\n",
    "    \n",
    "plt.savefig(f'{outDataDir}/figures/R1R2_Scaling_Combo.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot scaling and slope separately\n",
    "sns.set_style(\"ticks\")\n",
    "sns.set_context(\"paper\")\n",
    "\n",
    "f, ax = plt.subplots(\n",
    "    figsize=(2, 2))\n",
    "\n",
    "for cond in SepConds:\n",
    "    ax.loglog(\n",
    "        log_exp_agg[cond][0]['dist.avg']*1000,\n",
    "        log_exp_agg[cond][0]['balanced.avg'],\n",
    "        color = sampleColors[cond],\n",
    "        label = samplePlotNames[cond],\n",
    "        linestyle = sampleLineStyles[cond]\n",
    "    )\n",
    "\n",
    "    ax.set(\n",
    "        xlabel='separation, bp',\n",
    "        ylabel='IC contact frequency',\n",
    "        xlim=(1e3,1e8)\n",
    "    )\n",
    "    #ax.set_aspect(1.0)\n",
    "    ax.grid(lw=0.5)\n",
    "    ax.legend(bbox_to_anchor=(1.04,1), loc=\"upper left\", frameon = False)\n",
    "    \n",
    "plt.savefig(f'{outDataDir}/figures/R1R2_Scaling_Separate.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot slope - combo\n",
    "sns.set_style(\"ticks\")\n",
    "sns.set_context(\"paper\")\n",
    "\n",
    "f, ax = plt.subplots(\n",
    "    figsize=(2, 2))\n",
    "\n",
    "for cond in ComboConds:\n",
    "    ax.semilogx(\n",
    "        log_exp_agg[cond][1]['dist.avg']*1000,\n",
    "        log_exp_agg[cond][1]['slope'],\n",
    "        color = sampleColors[cond],\n",
    "        label = samplePlotNames[cond],\n",
    "        ls = sampleLineStyles[cond],\n",
    "        lw = 2\n",
    "    )\n",
    "\n",
    "ax.set(\n",
    "    xlabel='separation (s), bp',\n",
    "    ylabel='P(s) slope',\n",
    "    ylim=(-2.5, 0),\n",
    "    xlim=(1e3,1e8)\n",
    "    )\n",
    "\n",
    "ax.grid(lw=0.5)\n",
    "ax.legend(bbox_to_anchor=(1.04,1), loc=\"upper left\", frameon = False)\n",
    "plt.title('P(s) Slope')\n",
    "plt.axvline(x=1.5e7)\n",
    "    \n",
    "plt.savefig(f'{outDataDir}/figures/ScalingSlope_Combo_With1.5e7vline.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot slope - combo\n",
    "sns.set_style(\"ticks\")\n",
    "sns.set_context(\"paper\")\n",
    "\n",
    "f, ax = plt.subplots(\n",
    "    figsize=(2, 2))\n",
    "\n",
    "for cond in ComboConds:\n",
    "    ax.semilogx(\n",
    "        log_exp_agg[cond][1]['dist.avg']*1000,\n",
    "        log_exp_agg[cond][1]['slope'],\n",
    "        color = sampleColors[cond],\n",
    "        label = samplePlotNames[cond],\n",
    "        ls = sampleLineStyles[cond],\n",
    "        lw = 2\n",
    "    )\n",
    "\n",
    "ax.set(\n",
    "    xlabel='separation (s), bp',\n",
    "    ylabel='P(s) slope',\n",
    "    ylim=(-2.5, 0),\n",
    "    xlim=(1e3,1e8)\n",
    "    )\n",
    "\n",
    "ax.grid(lw=0.5)\n",
    "ax.legend(bbox_to_anchor=(1.04,1), loc=\"upper left\", frameon = False)\n",
    "\n",
    "plt.title('P(s) Slope')\n",
    "    \n",
    "plt.savefig(f'{outDataDir}/figures/ScalingSlope_Combo.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot scaling and slope separately\n",
    "sns.set_style(\"ticks\")\n",
    "sns.set_context(\"paper\")\n",
    "\n",
    "f, ax = plt.subplots(\n",
    "    figsize=(2, 2))\n",
    "\n",
    "for cond in SepConds:\n",
    "    ax.semilogx(\n",
    "        log_exp_agg[cond][1]['dist.avg']*1000,\n",
    "        log_exp_agg[cond][1]['slope'],\n",
    "        color = sampleColors[cond],\n",
    "        label = samplePlotNames[cond],\n",
    "        linestyle = sampleLineStyles[cond]\n",
    "    )\n",
    "\n",
    "    ax.set(\n",
    "        xlabel='separation, bp',\n",
    "        ylabel='slope',\n",
    "        ylim=(-2.2, 0),\n",
    "        xlim=(1e3,1e8)\n",
    "    )\n",
    "\n",
    "    ax.grid(lw=0.5)\n",
    "    #ax.set_aspect(1)\n",
    "    ax.legend(bbox_to_anchor=(1.04,1), loc=\"upper left\", frameon = False)\n",
    "    \n",
    "plt.savefig(f'{outDataDir}/figures/R1R2_ScalingSlope_Separate.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Call insulation, pileup at control boundaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#coolers - 10kb bins\n",
    "binsize = 10000\n",
    "\n",
    "clr_paths_10kb = {}\n",
    "for cond in conditions:\n",
    "    clr_paths_10kb[cond] = f'{outDataDir}/data/{long_names[cond]}.sampled.hg38.mapq_30.1000.mcool::resolutions/{binsize}'\n",
    "    \n",
    "clrs10kb = {\n",
    "    cond: cooler.Cooler(clr_paths_10kb[cond]) for cond in conditions\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use bioframe to fetch the genomic features from the UCSC.\n",
    "hg38_chromsizes = bioframe.fetch_chromsizes('hg38')\n",
    "hg38_cens = bioframe.fetch_centromeres('hg38')\n",
    "# create a view with chromosome arms using chromosome sizes and definition of centromeres\n",
    "hg38_arms = bioframe.make_chromarms(hg38_chromsizes,  hg38_cens)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select only chromosomes that are present in the good chromosomes\n",
    "hg38_chromsizes = hg38_chromsizes.loc[good_chroms]\n",
    "hg38_arms = hg38_arms[hg38_arms.chrom.isin(good_chroms)].reset_index(drop=True)\n",
    "hg38_chrom_vf = bioframe.make_viewframe(hg38_chromsizes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#call insulation\n",
    "insulation_table = {}\n",
    "\n",
    "for cond in conditions:\n",
    "    insulation_table[cond] = insulation(clrs10kb[cond], 250000, verbose=False, ignore_diags = 2, view_df=hg38_chrom_vf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "insulation_table[cond]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Functions to help with plotting\n",
    "def pcolormesh_45deg(ax, matrix_c, start=0, resolution=1, *args, **kwargs):\n",
    "    start_pos_vector = [start+resolution*i for i in range(len(matrix_c)+1)]\n",
    "    import itertools\n",
    "    n = matrix_c.shape[0]\n",
    "    t = np.array([[1, 0.5], [-1, 0.5]])\n",
    "    matrix_a = np.dot(np.array([(i[1], i[0])\n",
    "                                for i in itertools.product(start_pos_vector[::-1],\n",
    "                                                           start_pos_vector)]), t)\n",
    "    x = matrix_a[:, 1].reshape(n + 1, n + 1)\n",
    "    y = matrix_a[:, 0].reshape(n + 1, n + 1)\n",
    "    im = ax.pcolormesh(x, y, np.flipud(matrix_c), *args, **kwargs)\n",
    "    im.set_rasterized(True)\n",
    "    return im\n",
    "\n",
    "from matplotlib.ticker import EngFormatter\n",
    "bp_formatter = EngFormatter('b')\n",
    "def format_ticks(ax, x=True, y=True, rotate=True):\n",
    "    if y:\n",
    "        ax.yaxis.set_major_formatter(bp_formatter)\n",
    "    if x:\n",
    "        ax.xaxis.set_major_formatter(bp_formatter)\n",
    "        ax.xaxis.tick_bottom()\n",
    "    if rotate:\n",
    "        ax.tick_params(axis='x',rotation=45)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.colors import LogNorm\n",
    "from mpl_toolkits.axes_grid import make_axes_locatable\n",
    "import bioframe\n",
    "\n",
    "sns.set_style(\"ticks\")\n",
    "sns.set_context(\"paper\")\n",
    "\n",
    "fig = plt.figure(figsize=(8, 5))\n",
    "\n",
    "gs0 = GridSpec(nrows = 5, ncols = 2, width_ratios = [50, 1], hspace = 0.5)\n",
    "\n",
    "for i, cond in enumerate(ComboConds):\n",
    "    ax = plt.subplot(gs0[i, 0])\n",
    "    region = ('chr14', 70000000, 80000000)\n",
    "    norm = LogNorm(vmax=0.1, vmin=0.001)\n",
    "    data = clrs10kb[cond].matrix(balance=True).fetch(region)\n",
    "    im = pcolormesh_45deg(ax, data, start=region[1], resolution=10000, norm=norm, cmap='fall')\n",
    "    ax.set_ylim(0, 2000000)\n",
    "    format_ticks(ax, rotate=False)\n",
    "    ax.xaxis.set_visible(False)\n",
    "    ax.set_xlim(region[1], region[2])\n",
    "    plt.title(samplePlotNames[cond])\n",
    "    plt.ylabel('Distance')\n",
    "\n",
    "    #divider = make_axes_locatable(ax)\n",
    "    cax = plt.subplot(gs0[i, 1])\n",
    "    plt.colorbar(im, cax=cax)\n",
    "    cax.set_aspect(6)\n",
    "\n",
    "#ins_ax = divider.append_axes(\"bottom\", size=\"50%\", pad=0., sharex=ax)\n",
    "    \n",
    "for cond in ComboConds:\n",
    "    ins_ax = plt.subplot(gs0[4, 0])\n",
    "    insul_region = bioframe.select(insulation_table[cond], region)\n",
    "    ins_ax.plot(insul_region[['start', 'end']].mean(axis=1), \n",
    "                insul_region['log2_insulation_score_250000'], \n",
    "                label=samplePlotNames[cond], color = sampleColors[cond])\n",
    "    plt.ylim(-2.25, 1.25)\n",
    "    format_ticks(ins_ax, y=False, rotate=False)\n",
    "    ins_ax.legend().set_visible(False)\n",
    "    ins_ax.set_xlim(region[1], region[2])\n",
    "    plt.ylabel('Insulation')\n",
    "    plt.xlabel('Chr14 Coordinates')\n",
    "    \n",
    "    h, l = ins_ax.get_legend_handles_labels() # get labels and handles\n",
    "    leg_ax = plt.subplot(gs0[4, 1])\n",
    "    leg_ax.legend(h, l, frameon = False, bbox_to_anchor=(0, 1), loc=\"upper left\")  \n",
    "    leg_ax.axis('off')\n",
    "    \n",
    "plt.savefig(f'{outDataDir}/figures/All_insulation_example.png', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from skimage.filters import threshold_li, threshold_otsu\n",
    "windows = [250000]\n",
    "histkwargs = dict(\n",
    "    bins=10**np.linspace(-4,1,200),\n",
    "    histtype='step',\n",
    "    lw=2,\n",
    ")\n",
    "\n",
    "fig = plt.figure(figsize=(12, 16))\n",
    "\n",
    "gs0 = GridSpec(nrows = 4, ncols = 2, wspace=0.3, hspace = 0.3)\n",
    "\n",
    "thresholds_li = {}\n",
    "thresholds_otsu = {}\n",
    "bounds_li = {}\n",
    "bounds_otsu = {}\n",
    "\n",
    "for i, cond in enumerate(ComboConds):\n",
    "    gs1 = GridSpecFromSubplotSpec(nrows=len(windows), ncols=1, wspace = 0.3, hspace = 1, subplot_spec = gs0[i])\n",
    "\n",
    "    thresholds_li[cond] = {}\n",
    "    thresholds_otsu[cond] = {}\n",
    "    bounds_li[cond] = {}\n",
    "    bounds_otsu[cond] = {}\n",
    "\n",
    "    for i, w in enumerate(windows):\n",
    "        ax = plt.subplot(gs1[i])\n",
    "        ax.hist(\n",
    "            insulation_table[cond][f'boundary_strength_{w}'],\n",
    "            **histkwargs\n",
    "        )\n",
    "        thresholds_li[cond][w] = threshold_li(insulation_table[cond][f'boundary_strength_{w}'].dropna().values)\n",
    "        thresholds_otsu[cond][w] = threshold_otsu(insulation_table[cond][f'boundary_strength_{w}'].dropna().values)\n",
    "        n_boundaries_li = (insulation_table[cond][f'boundary_strength_{w}'].dropna()>=thresholds_li[cond][w]).sum()\n",
    "        bounds_li[cond][w] = insulation_table[cond][(insulation_table[cond][f'boundary_strength_{w}']>=thresholds_li[cond][w])]\n",
    "        n_boundaries_otsu = (insulation_table[cond][f'boundary_strength_{w}'].dropna()>=thresholds_otsu[cond][w]).sum()\n",
    "        bounds_otsu[cond][w] = insulation_table[cond][(insulation_table[cond][f'boundary_strength_{w}']>=thresholds_otsu[cond][w])]\n",
    "\n",
    "        ax.axvline(thresholds_li[cond][w], c='green')\n",
    "        ax.axvline(thresholds_otsu[cond][w], c='magenta')\n",
    "        ax.text(0.01, 0.9,\n",
    "                 f'Window {w//1000}kb',\n",
    "                 ha='left',\n",
    "                 va='top',\n",
    "                 transform=ax.transAxes)\n",
    "        ax.text(0.01, 0.7,\n",
    "                f'{n_boundaries_otsu} boundaries (Otsu)',\n",
    "                c='magenta',\n",
    "                ha='left',\n",
    "                va='top',\n",
    "                transform=ax.transAxes)\n",
    "        ax.text(0.01, 0.5,\n",
    "                f'{n_boundaries_li} boundaries (Li)',\n",
    "                c='green',\n",
    "                ha='left',\n",
    "                va='top',\n",
    "                transform=ax.transAxes)\n",
    "\n",
    "        ax.set(\n",
    "            xscale='log',\n",
    "            ylabel='# boundaries'\n",
    "        )\n",
    "\n",
    "        ax.set(xlabel='Boundary strength')\n",
    "        plt.title(cond)\n",
    "    \n",
    "plt.savefig(f'{outDataDir}/figures/AllCond_BoundaryThresholds.png', dpi = 300, \n",
    "            bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use bioframe to fetch the genomic features from the UCSC.\n",
    "hg38_chromsizes = bioframe.fetch_chromsizes('hg38')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hg38_chromsizes = hg38_chromsizes.loc[good_chroms]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for cond in conditions:\n",
    "    for w in windows:\n",
    "        bioframe.to_bigwig(insulation_table[cond], hg38_chromsizes,\n",
    "            f'{outDataDir}/data/{long_names[cond]}.{binsize//1000}kb.mapq30.insul_score_{w}.bw', \n",
    "            f'log2_insulation_score_{w}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import bbi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create stackup. flank = .5Mb, nbins = 100, 250kb window, 10kb hi-c resolution\n",
    "flank = 500000 # Length of flank to one side from the boundary, in basepairs\n",
    "nbins = 100   # Number of bins to split the region\n",
    "ins_pileup_signal = {}\n",
    "for cond in conditions:\n",
    "    ins_pileup_signal[cond] = bbi.stackup(\n",
    "        f'{outDataDir}/data/{long_names[cond]}.{binsize//1000}kb.mapq30.insul_score_250000.bw',\n",
    "        bounds_otsu['WT_Ctrl_R1R2'][250000].chrom,\n",
    "        bounds_otsu['WT_Ctrl_R1R2'][250000].start-flank,\n",
    "        bounds_otsu['WT_Ctrl_R1R2'][250000].end+flank, bins=nbins)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"ticks\")\n",
    "sns.set_context(\"paper\")\n",
    "\n",
    "f, ax = plt.subplots(figsize=[2, 2])\n",
    "flank = 500000\n",
    "nbins = 100\n",
    "x = np.linspace(-flank/1e3, flank/1e3, nbins)\n",
    "for cond in SepConds:\n",
    "    ax.plot(x, np.nanmean(ins_pileup_signal[cond], axis=0), color = sampleColors[cond], label = cond, \n",
    "            ls = sampleLineStyles[cond], lw = 2)\n",
    "plt.xlim(-flank/1e3, flank/1e3)\n",
    "plt.ylim(-1, .3)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0.1, frameon = False)\n",
    "plt.title('Insulation Profile \\nWT Ctrl (R1+R2) Boundaries')\n",
    "plt.ylabel('Average Insulation')\n",
    "plt.xlabel('Distance from Boundary (Kb)')\n",
    "plt.savefig(f'{outDataDir}/figures/InsulationPileup_WTCtrlR1R2Boundaries.png', dpi = 300, \n",
    "            bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#separate out TAD vs Compartment boundaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#called boundaries from cooler compartments (with gene density)\n",
    "binsize_comp = 250000\n",
    "\n",
    "eigs = {}\n",
    "for cond in conditions:\n",
    "    eigs[cond] = pd.read_csv(f'{outDataDir}/data/{long_names[cond]}.{binsize_comp//1000}kb.mapq30.byarm.eigs.cis.vecs.txt', sep='\\t')\n",
    "\n",
    "#calling A vs B compartments in each sample\n",
    "eigsA = {}\n",
    "for cond in conditions:\n",
    "    eigsA[cond] = eigs[cond][eigs[cond]['E1'] > 0]\n",
    "    \n",
    "eigsB = {}\n",
    "for cond in conditions:\n",
    "    eigsB[cond] = eigs[cond][eigs[cond]['E1'] < 0]\n",
    "\n",
    "compRangesA = {}\n",
    "for cond in conditions:\n",
    "    compRangesA[cond] = bioframe.merge(pd.DataFrame(data = {\n",
    "        'chrom' : eigsA[cond].chrom,\n",
    "        'start' : eigsA[cond].start,\n",
    "        'end' : eigsA[cond].end,\n",
    "        'name' : '.',\n",
    "        'score' : '.',\n",
    "        'strand' : '+'\n",
    "    }))\n",
    "    \n",
    "compRangesB = {}\n",
    "for cond in conditions:\n",
    "    compRangesB[cond] = bioframe.merge(pd.DataFrame(data = {\n",
    "        'chrom' : eigsB[cond].chrom,\n",
    "        'start' : eigsB[cond].start,\n",
    "        'end' : eigsB[cond].end,\n",
    "        'name' : '.',\n",
    "        'score' : '.',\n",
    "        'strand' : '+'\n",
    "    }))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calling compartment boundaries in each sample\n",
    "#Calculate which bins flank sign changes - on same chromosome. \n",
    "\n",
    "compBounds = {}\n",
    "signsAll = {}\n",
    "\n",
    "for cond in conditions:    \n",
    "    signsAll[cond] = pd.DataFrame(data = {\n",
    "        'Chrom1' : eigs[cond]['chrom'][0:-1].reset_index(drop = True),\n",
    "        'Bin1Start' : eigs[cond]['start'][0:-1].reset_index(drop = True),\n",
    "        'Bin1End' : eigs[cond]['end'][0:-1].reset_index(drop = True),\n",
    "        'Bin1E1' : eigs[cond]['E1'][0:-1].reset_index(drop = True),\n",
    "        'Chrom2' : eigs[cond]['chrom'][1:].reset_index(drop = True),\n",
    "        'Bin2Start' : eigs[cond]['start'][1:].reset_index(drop = True),\n",
    "        'Bin2End' : eigs[cond]['end'][1:].reset_index(drop = True),  \n",
    "        'Bin2E1' : eigs[cond]['E1'][1:].reset_index(drop = True),\n",
    "        'sign' : (eigs[cond]['E1'][0:-1].reset_index(drop = True) * eigs[cond]['E1'][1:].reset_index(drop = True))\n",
    "        })\n",
    "    signsAll[cond].apply(pd.to_numeric, errors='ignore')\n",
    "    signsAll[cond] = signsAll[cond].astype({\"Bin1Start\": int, \"Bin1End\": int, \"Bin2Start\": int, \"Bin2End\": int})\n",
    "\n",
    "    compBounds[cond] = signsAll[cond][(signsAll[cond]['sign'] < 0) &\n",
    "                                      (signsAll[cond]['Chrom1'] == signsAll[cond]['Chrom2']) \n",
    "                                     ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "compBoundsRanges = {}\n",
    "for cond in conditions:\n",
    "    compBoundsRanges[cond] = pd.DataFrame(data = {\n",
    "        'chrom' : compBounds[cond].Chrom1,\n",
    "        'start' : compBounds[cond].Bin1Start,\n",
    "        'end' : compBounds[cond].Bin2End,\n",
    "        'name' : '.',\n",
    "        'score' : compBounds[cond].sign,\n",
    "        'strand' : '+'\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "compBoundsRanges['WT_Ctrl_R1R2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bounds_otsu[cond][250000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tadOnlyBoundaries = {}\n",
    "overlapCompBoundaries = {}\n",
    "\n",
    "for cond in ComboConds:\n",
    "    insBoundaries = bounds_otsu[cond][250000].copy()\n",
    "    compBoundaries = compBoundsRanges[cond].copy()\n",
    "    overlapBounds = bioframe.overlap(insBoundaries, compBoundaries, how = 'inner').iloc[:, 0:4]\n",
    "    overlapBounds.columns = ['chrom', 'start', 'end', 'boundary_strength_250000']\n",
    "    overlapCompBoundaries[cond] = overlapBounds.drop_duplicates()\n",
    "    tadOnlyBoundaries[cond] = bioframe.subtract(insBoundaries, overlapCompBoundaries[cond]).drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save the comp vs tad insulation boundaries, and all insulation boundaries\n",
    "for cond in ComboConds:\n",
    "    overlapCompBoundaries[cond].to_csv(\n",
    "        f'{outDataDir}/data/{long_names[cond]}.compartmentOverlap_insulation_250kbWindow_bounds.mapq30.goodchroms.arms.txt',\n",
    "        sep = \"\\t\", index = False)\n",
    "    tadOnlyBoundaries[cond].to_csv(    \n",
    "        f'{outDataDir}/data/{long_names[cond]}.TADonly_insulation_250kbWindow_bounds.mapq30.goodchroms.arms.txt',\n",
    "        sep = \"\\t\", index = False)\n",
    "    bounds_otsu[cond][250000].to_csv(    \n",
    "        f'{outDataDir}/data/{long_names[cond]}.AllBounds_insulation_250kbWindow_bounds.mapq30.goodchroms.arms.txt',\n",
    "        sep = \"\\t\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in comp vs tad insulation boundaries\n",
    "tadOnlyBoundaries = {}\n",
    "overlapCompBoundaries = {}\n",
    "\n",
    "for cond in ComboConds:\n",
    "    overlapCompBoundaries[cond] = pd.read_csv(\n",
    "        f'{outDataDir}/data/{long_names[cond]}.compartmentOverlap_insulation_250kbWindow_bounds.mapq30.goodchroms.arms.txt',\n",
    "        sep = \"\\t\")\n",
    "    tadOnlyBoundaries[cond] = pd.read_csv(    \n",
    "        f'{outDataDir}/data/{long_names[cond]}.TADonly_insulation_250kbWindow_bounds.mapq30.goodchroms.arms.txt',\n",
    "        sep = \"\\t\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create stackup. flank = .5Mb, nbins = 100, 250kb window, 10kb hi-c resolution, TAD only boundaries\n",
    "flank = 500000 # Length of flank to one side from the boundary, in basepairs\n",
    "nbins = 100   # Number of bins to split the region\n",
    "ins_pileup_signal_TAD = {}\n",
    "for cond in conditions:\n",
    "    ins_pileup_signal_TAD[cond] = bbi.stackup(\n",
    "        f'{outDataDir}/data/{long_names[cond]}.{binsize//1000}kb.mapq30.insul_score_250000.bw',\n",
    "        tadOnlyBoundaries['WT_Ctrl_R1R2'].chrom,\n",
    "        tadOnlyBoundaries['WT_Ctrl_R1R2'].start-flank,\n",
    "        tadOnlyBoundaries['WT_Ctrl_R1R2'].end+flank, bins=nbins)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"ticks\")\n",
    "sns.set_context(\"paper\")\n",
    "f, ax = plt.subplots(figsize=[2, 2])\n",
    "flank = 500000\n",
    "nbins = 100\n",
    "x = np.linspace(-flank/1e3, flank/1e3, nbins)\n",
    "for cond in SepConds:\n",
    "    ax.plot(x, np.nanmean(ins_pileup_signal_TAD[cond], axis=0), color = sampleColors[cond], label = samplePlotNames[cond], \n",
    "            ls = sampleLineStyles[cond], lw = 2)\n",
    "plt.xlim(-flank/1e3, flank/1e3)\n",
    "plt.ylim(-.8, .3)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0.1, frameon = False)\n",
    "plt.title('Insulation Profile \\n WT Ctrl Boundaries')\n",
    "plt.ylabel('Average Insulation')\n",
    "plt.xlabel('Distance from Boundary (Kb)')\n",
    "plt.savefig(f'{outDataDir}/figures/InsulationPileup_WTCtrlR1R2TADBoundaries.png', dpi = 300, \n",
    "            bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_style(\"ticks\")\n",
    "sns.set_context(\"paper\")\n",
    "f, ax = plt.subplots(figsize=[2, 2])\n",
    "flank = 500000\n",
    "nbins = 100\n",
    "x = np.linspace(-flank/1e3, flank/1e3, nbins)\n",
    "for cond in ComboConds:\n",
    "    ax.plot(x, np.nanmean(ins_pileup_signal_TAD[cond], axis=0), color = sampleColors[cond], label = samplePlotNames[cond], \n",
    "            ls = sampleLineStyles[cond], lw = 2)\n",
    "plt.xlim(-flank/1e3, flank/1e3)\n",
    "plt.ylim(-.8, .3)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0.1, frameon = False)\n",
    "plt.title('Insulation Profile \\n WT Ctrl TAD Boundaries')\n",
    "plt.ylabel('Average Insulation')\n",
    "plt.xlabel('Distance from Boundary (Kb)')\n",
    "plt.savefig(f'{outDataDir}/figures/InsulationPileup_WTCtrlR1R2TADBoundaries_Combo.png', dpi = 300, \n",
    "            bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#1kb bins - coverage in cis vs total"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#coolers - 1kb bins\n",
    "binsize = 1000\n",
    "\n",
    "clr_paths_1kb = {}\n",
    "for cond in conditions:\n",
    "    clr_paths_1kb[cond] = f'{outDataDir}/data/{long_names[cond]}.sampled.hg38.mapq_30.1000.mcool::resolutions/{binsize}'\n",
    "    \n",
    "clrs1kb = {\n",
    "    cond: cooler.Cooler(clr_paths_1kb[cond]) for cond in conditions\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clrs1kb['WT_Ctrl_R1'].info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ciscov = {}\n",
    "totcov = {}\n",
    "for cond in conditions:\n",
    "    ciscov[cond], totcov[cond] = cooltools.api.coverage.coverage(\n",
    "        clrs1kb[cond], \n",
    "        ignore_diags=2, \n",
    "        chunksize=10000000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#total sum is not the same as sum/count from cooler unless don't ignore 2 diags, but then cis % is not right (since not balanced)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ciscov[cond]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cis/trans - all chromosomes, ignore first 2 diags, from coolers\n",
    "cisratio = {}\n",
    "for cond in conditions:\n",
    "    cisratio[cond] = sum(ciscov[cond])/(sum(totcov[cond])-sum(ciscov[cond]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot bargraph of avg and dots of replicates for cis ratio\n",
    "\n",
    "cis_ratio_df = pd.DataFrame(columns = ['Condition', 'Replicate', 'Label', 'Ratio'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for cond in SepConds:\n",
    "    cis_ratio_df = cis_ratio_df.append({\n",
    "                'Condition' : f'{cond}', \n",
    "                'Replicate' : repdict[cond], \n",
    "                'Label' : f'{labeldict[cond]}',\n",
    "                'Ratio' : cisratio[cond]\n",
    "            }, ignore_index = True)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cis_ratio_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#https://stackoverflow.com/questions/64223870/seaborn-overlap-swarmplot-on-barplot\n",
    "sns.set_style(\"ticks\")\n",
    "sns.set_context(\"paper\")\n",
    "cmap_bar = sns.color_palette(['#a6cee3', '#1f78b4', '#b2df8a', '#33a02c'])\n",
    "gs = GridSpec(nrows= 1, ncols=1, wspace = 0.6, hspace = 0.6)\n",
    "\n",
    "plt.figure(figsize=(2, 2))\n",
    "\n",
    "ax = plt.subplot(gs[0])\n",
    "\n",
    "sns.stripplot(\n",
    "    x='Label',\n",
    "    y='Ratio',\n",
    "    hue='Label', \n",
    "    dodge=False, \n",
    "    data=cis_ratio_df, \n",
    "    jitter = False, \n",
    "    palette = cmap_bar, \n",
    "    ax = ax)\n",
    "ax1 = sns.barplot(\n",
    "    x='Label', \n",
    "    y='Ratio', \n",
    "    hue='Label', \n",
    "    data=cis_ratio_df,\n",
    "    palette = cmap_bar, \n",
    "    alpha = 0.5, \n",
    "    dodge = False,\n",
    "    ci = False, \n",
    "    ax = ax)\n",
    "    \n",
    "handles, labels = ax1.get_legend_handles_labels()\n",
    "ax1.legend(handles[4:8], labels[4:8], bbox_to_anchor=(1.04,1), frameon = False)\n",
    " \n",
    "plt.title(f'Cis/Trans Ratio')\n",
    "plt.ylabel('Cis/Trans Contacts')  \n",
    "#plt.ylim(0, 1)\n",
    "plt.xlabel('Sample')\n",
    "plt.xticks(rotation=90)\n",
    "        \n",
    "plt.savefig(f'{outDataDir}/figures/R1R2_HiC_CisvsTransRatio_BarWithScatter.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cis/trans - all chromosomes, ignore first 2 diags, from coolers\n",
    "cisratio = {}\n",
    "for cond in conditions:\n",
    "    cisratio[cond] = sum(ciscov[cond])/sum(totcov[cond])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot bargraph of avg and dots of replicates for cis ratio\n",
    "\n",
    "cis_ratio_df = pd.DataFrame(columns = ['Condition', 'Replicate', 'Label', 'Ratio'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for cond in SepConds:\n",
    "    cis_ratio_df = cis_ratio_df.append({\n",
    "                'Condition' : f'{cond}', \n",
    "                'Replicate' : repdict[cond], \n",
    "                'Label' : f'{labeldict[cond]}',\n",
    "                'Ratio' : cisratio[cond]\n",
    "            }, ignore_index = True)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cis_ratio_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#https://stackoverflow.com/questions/64223870/seaborn-overlap-swarmplot-on-barplot\n",
    "sns.set_style(\"ticks\")\n",
    "sns.set_context(\"paper\")\n",
    "cmap_bar = sns.color_palette(['#a6cee3', '#1f78b4', '#b2df8a', '#33a02c'])\n",
    "gs = GridSpec(nrows= 1, ncols=1, wspace = 0.6, hspace = 0.6)\n",
    "\n",
    "plt.figure(figsize=(2, 2))\n",
    "\n",
    "ax = plt.subplot(gs[0])\n",
    "\n",
    "sns.stripplot(\n",
    "    x='Label',\n",
    "    y='Ratio',\n",
    "    hue='Label', \n",
    "    dodge=False, \n",
    "    data=cis_ratio_df, \n",
    "    jitter = False, \n",
    "    palette = cmap_bar, \n",
    "    ax = ax)\n",
    "ax1 = sns.barplot(\n",
    "    x='Label', \n",
    "    y='Ratio', \n",
    "    hue='Label', \n",
    "    data=cis_ratio_df,\n",
    "    palette = cmap_bar, \n",
    "    alpha = 0.5, \n",
    "    dodge = False,\n",
    "    ci = False, \n",
    "    ax = ax)\n",
    "    \n",
    "handles, labels = ax1.get_legend_handles_labels()\n",
    "ax1.legend(handles[4:8], labels[4:8], bbox_to_anchor=(1.04,1), frameon = False)\n",
    " \n",
    "plt.title(f'Fraction Intrachromosomal')\n",
    "plt.ylabel('Cis/Total Contacts')  \n",
    "plt.ylim(0.8, 0.9)\n",
    "plt.xlabel('Sample')\n",
    "plt.xticks(rotation=90)\n",
    "        \n",
    "plt.savefig(f'{outDataDir}/figures/R1R2_HiC_CisvsTotalRatio_BarWithScatter.png', dpi = 300, bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
